### DISCUSSION TITLE
Using eleventy-fetch in multiple javascript data files

### DISCUSSION THREAD

**Discussion by danielLee0721** [11ty/eleventy#2614] [Q&A] [answered]
Hi everyone,

I am not sure if I am missing something but when I use the eleventy-fetch plugin (https://www.11ty.dev/docs/plugins/fetch/) in a single javascript file, let's say a global javascript data file, then everything works as expected but when I use it in multiple places (global javascript data file and some template javascript data files), then it doesn't fetch data for all of them. Does this mean that eleventy-fetch-plugin can only be used in one data file? Is this normal behaviour?

Thank you

**Comment by pdehaan** [comment, marked as answer]
It should work in multiple places (assuming you're always using it in an `async` context).
But you would probably need to share some code or a repo for a more helpful answer since it's hard to troubleshoot.

And you might get a better response in the https://github.com/11ty/eleventy-fetch/issues repo.

**Reply by danielLee0721** [reply]
Thank you for the response. I was able to get the elevety-fetch plugin work in multiple data files.

I now have a follow up question.

If I have

src/company/users.html
```
---
layout: defaultPage.html
---
```


src/company/users.11tydata.js
```
const EleventyFetch = require('@11ty/eleventy-fetch');

module.exports = async function () {
  let url = 'https://jsonplaceholder.typicode.com/users';

  let userData = await EleventyFetch(url, {
    duration: '2s',
    type: 'json'
  });

  return userData;
};
```
and this returns an array of users.

How can I loop through this array of data and output the data? I am not sure how to refer to the array...


Thank you

**Reply by pdehaan** [reply]
I'd make one minor tweak, and make sure your users.11ty.js data file returns an object instead of an array:

```diff
// src/company/users.11tydata.js

const EleventyFetch = require('@11ty/eleventy-fetch');

module.exports = async function () {
  let url = 'https://jsonplaceholder.typicode.com/users';

  let userData = await EleventyFetch(url, {
    duration: '2s',
    type: 'json'
  });
-  return userData;
+  return { userData };
};
```

Now, in your users.html, you can use a LiquidJS `{% for %}` loop to iterate over the array of users [by referring to our `userData` array exported from the data file]:

```liquid
---
title: Users
---

<h1>{{ title }}</h1>
{% for user in userData %}
<article>
  <h2 data-id="{{ user.id }}">{{ user.name }}</h2>
  <p>email={{ user.email }}</p>
  <p>phone={{ user.phone }}</p>
  <p>website={{ user.website }}</p>
</article>
{% endfor %}
```

### OUTPUT

```html
<h1>Users</h1>

<article>
  <h2 data-id="1">Leanne Graham</h2>
  <p>email=Sincere@april.biz</p>
  <p>phone=1-770-736-8031 x56442</p>
  <p>website=hildegard.org</p>
</article>

<article>
  <h2 data-id="2">Ervin Howell</h2>
  <p>email=Shanna@melissa.tv</p>
  <p>phone=010-692-6593 x09125</p>
  <p>website=anastasia.net</p>
</article>

...
```

**Reply by danielLee0721** [reply]
Hi Peter,

Thank you for your help.

One quick question - how did you show git diff like changes in your reply?
```
-  return userData;
+  return { userData };
```

Thank you!

**Reply by pdehaan** [reply]
I just set my code fence language to "diff" to make it easier to read:

````
```diff
 one
- two
+ three
```
````

```diff
 one
- two
+ three
```

**Reply by danielLee0721** [reply]
Hi Peter - sorry to bother you again but I seem to be having problems again.. I invited you to a private repo for your kind review. Thank you very much in advance.

Basically I am trying to fetch data in three different files.

One global data file
_data/globalData.js

and two template data files

When I comment out any two out of the three files then everything works as expected but if I have more than one file active then it only fetches one of the files.

**Reply by pdehaan** [reply]
Not sure why that'd happen&hellip;
I might need to try cloning locally and replacing the GraphQL queries to POST queries from some other public API and see if I can reproduce.

**Reply by pdehaan** [reply]
Simple proof of concept worked for me (fetching in a global data file and 2 template data files).

Here's my sample repo: https://github.com/pdehaan/11ty-2614-2

```sh
tree src
src
├── _data
│   └── globalData.js
├── albums
│   ├── index.11tydata.js
│   └── index.html
├── index.html
└── photos
    ├── index.11tydata.js
    └── index.html

3 directories, 6 files
```

### src/_data/globalData.js
```js
const EleventyFetch = require('@11ty/eleventy-fetch');

module.exports = async function () {
  let url = 'https://jsonplaceholder.typicode.com/todos/1';
  let globalData = await EleventyFetch(url, {
    fetchOptions: { },
    duration: '30s',
    type: 'json'
  });

  return globalData;
};
```

### src/albums/index.11tydata.js
```js
const EleventyFetch = require('@11ty/eleventy-fetch');

module.exports = async function () {
  const url = 'https://jsonplaceholder.typicode.com/albums';
  const albums = await EleventyFetch(url, {
    fetchOptions: {},
    duration: '30s',
    type: 'json'
  });
  return { albums };
};
```

### src/photos/index.11tydata.js
```js
const EleventyFetch = require('@11ty/eleventy-fetch');

module.exports = async function () {
  const url = 'https://jsonplaceholder.typicode.com/photos';
  const photos = await EleventyFetch(url, {
    fetchOptions: {},
    duration: '30s',
    type: 'json'
  });
  return { photos };
};
```

**Reply by danielLee0721** [reply]
Very strange.. I think the problem has to do with DatoCMS.. I will reach out to them! Thank you so much.

**Reply by pdehaan** [reply]
If I remember correctly, you had fairly low caching numbers (1-2 seconds). So might be some quota issues.
You should also be able to see the contents of the cached data locally in your .cache folder to see if something is maybe undefined or an empty array/object.

Not sure what else to try. You could maybe remove 11ty from the equation and just call the three data files directly with some `console.log()` code (just above the final `return` statements) to see if Dato/11ty is reliably returning data.

Or a less popular solution of sending me your `process.env.DATO_API_TOKEN` and I can try running your code locally to see if I can reproduce your results -- on the off chance that it is something related to the GraphQL-ness that would be present in my simplified jsonplaceholder examples.

**Reply by danielLee0721** [reply]
Thank you for looking into this for me. The token is inside the .env in the private repo. It seems like the cached files are overwriting one another.

**Reply by pdehaan** [reply]
> _"It seems like the cached files are overwriting one another."_

Oh, yeah! I think we discussed this elsewhere, or maybe I'm crossing threads.
When doing POST requests, it seems to store the specified URL as the unique cache key. So in your case, all three requests get normalized to the same generic GraphQL POST URL which won't work.

I was able to "fix" this by using unique request hash URLs for each of the three instances:

```diff
-let url = 'https://graphql.datocms.com/';
+let url = 'https://graphql.datocms.com/#facelift';
```
```diff
-let url = 'https://graphql.datocms.com/';
+let url = 'https://graphql.datocms.com/#tummy-tuck';
```
```diff
-let url = 'https://graphql.datocms.com/';
+let url = 'https://graphql.datocms.com/#homepage';
```

This way eleventy-fetch will use the new "unique" URLs as caching keys instead of all three instances overwriting the generic "graphql.datocms.com" POST URL.

And then I think in your src/index.html, you had a typo where you need to use `{{ globalData.homePage.title }}` instead of just `{{ globalData.title }}` (or, you could also just modify the globalData response to return `globalData.data.homepage` instead of `globalData.data` if you prefer):

```diff
---
layout: defaultPage.html
title:
description:
htmlClass:
bodyClass:
---
-{{ globalData.title }}
+{{ globalData.homePage.title }}
```

**Reply by danielLee0721** [reply]
Works like a charm! Thank you very much. Always learning things from you. 🫡

And yes, thank you for pointing out the typo. 🙏🏼
